x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFunction.java

```diff
@@ -22,147 +20,192 @@
+abstract public class DateTimeFunction extends ScalarFunction {
 
-    private final TimeZone timeZone;
+    final static TimeZone UTC = TimeZone.getTimeZone("UTC");
+
+    private final FunctionContext context;
     private final String name;
 
-    DateTimeFunction(Location location, Expression field, TimeZone timeZone) {
-        super(location, field);
-        this.timeZone = timeZone;
+    DateTimeFunction(Location location, List<Expression> arguments, FunctionContext context) {
+        super(location, arguments);
+        this.context = context;
 
         StringBuilder sb = new StringBuilder(super.name());
         // add timezone as last argument
-        sb.insert(sb.length() - 1, " [" + timeZone.getID() + "]");
+        sb.insert(sb.length() - 1, " [" + context + " ]");
 
         this.name = sb.toString();
     }
 
     @Override
     protected final NodeInfo<DateTimeFunction> info() {
-        return NodeInfo.create(this, ctorForInfo(), field(), timeZone());
+        return NodeInfo.create(this, ctorForInfo(), arguments(), context());
     }
-    protected abstract NodeInfo.NodeCtor2<Expression, TimeZone, DateTimeFunction> ctorForInfo();
 
-    public TimeZone timeZone() {
-        return timeZone;
+    abstract NodeInfo.NodeCtor2<List<Expression>, FunctionContext, DateTimeFunction> ctorForInfo();
+
+    public final Expression field() {
+        return this.arguments().get(0);
     }
 
-    @Override
-    public boolean foldable() {
-        return field().foldable();
+    public final FunctionContext context() {
+        return context;
     }
 
     @Override
-    public Object fold() {
-        DateTime folded = (DateTime) field().fold();
-        if (folded == null) {
-            return null;
-        }
-
-        ZonedDateTime time = ZonedDateTime.ofInstant(
-            Instant.ofEpochMilli(folded.getMillis()), ZoneId.of(timeZone.getID()));
-        return time.get(chronoField());
+    public final Expression replaceChildren(List<Expression> newChildren) {
+        return this.replaceChildren(this.location(), newChildren, this.context());
     }
 
+    abstract Expression replaceChildren(Location location, List<Expression> newChildren, FunctionContext context);
+
     @Override
-    protected TypeResolution resolveType() {
-        if (field().dataType() == DataType.DATE) {
-            return TypeResolution.TYPE_RESOLVED;
-        }
-        return new TypeResolution("Function [" + functionName() + "] cannot be applied on a non-date expression (["
-                + Expressions.name(field()) + "] of type [" + field().dataType().esType + "])");
+    public final boolean foldable() {
+        return this.arguments().stream()
+            .map( e -> e.foldable())
+            .allMatch( b -> b);
     }
 
     @Override
-    protected ScriptTemplate asScriptFrom(FieldAttribute field) {
-        ParamsBuilder params = paramsBuilder();
+    public final Object fold() {
```

Comment:
big improvement now users can call date.getXXX with optional timezone and locale where applicable, instead of horrible boilerplate from before.